src/SqlBinding/TriggerBinding/SqlTableWatchers.cs

```diff
@@ -0,0 +1,1222 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.WebJobs.Host.Executors;
+using Microsoft.Data.SqlClient;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using SqlBinding.TriggerBinding;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using static SqlBinding.TriggerBinding.ScaleRecommendation;
+
+namespace Microsoft.Azure.WebJobs.Extensions.Sql
+{
+    public class SqlTableWatchers
+    {
+        private static string[] variableLengthTypes = new string[] { "varchar", "nvarchar", "nchar", "char", "binary", "varbinary" };
+        private static string[] variablePrecisionTypes = new string[] { "numeric", "decimal" };
+
+        public class SqlPerformanceMonitor
+        {
+            private string _workerTable;
+            private readonly string _globalStateTable;
+            private readonly string _userTable;
+            private readonly string _connectionString;
+            
+            private readonly Dictionary<string, string> _primaryKeys;
+            private string _leftOuterJoinWorkerTable;
+            private readonly ILogger _logger;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="SqlPerformanceMonitor"> class
+            /// </summary>
+            /// <param name="connectionString">
+            /// The SQL connection string used to connect to the user's database
+            /// </param>
+            /// <param name="table"> 
+            /// The name of the user table that changes are being tracked on
+            /// </param>
+            /// <exception cref="ArgumentException">
+            /// Thrown if table or connectionString are null or empty
+            /// </exception>
+            /// <exception cref="ArgumentNullException">
+            /// Thrown if the logger is null
+            /// </exception>
+            public SqlPerformanceMonitor(string table, string connectionString, ILogger logger)
+            {
+                if (string.IsNullOrEmpty(table))
+                {
+                    throw new ArgumentException("User table name cannot be null or empty");
+                }
+                if (string.IsNullOrEmpty(connectionString))
+                {
+                    throw new ArgumentException("SQL connection string cannot be null or empty");
+                }
+
+                _connectionString = connectionString;
+                _userTable = SqlBindingUtilities.NormalizeTableName(table);
+                _globalStateTable = "[az_func].[Global_State_Table]";
+                _primaryKeys = new Dictionary<string, string>();
```

Comment:
Does it make sense in this case, given that we're storing the primary key column names of the user table? It could be the case that the user table has two different columns, both part of the primary key, that are named Names and names. Would be weird, but is conceivably possible.